pandas/io/formats/format.py

```diff
@@ -656,6 +656,10 @@ def _chk_truncate(self) -> None:
                 frame = concat(
                     (frame.iloc[:, :col_num], frame.iloc[:, -col_num:]), axis=1
                 )
+                # truncate formatter
+                if is_list_like(self.formatters) and self.formatters:
+                    truncate_fmt = cast(List[Callable], self.formatters)
```

Comment:
Just out of curiosity what complain was this giving? cast doesn't seem necessary given assignment back to self.formatters but I may be missing something

Reply:
It's actually a mypy complain, when running ci/code_checks.sh, about the possibility of self.formatters not been a list.
<img width="1188" alt="image" src="https://user-images.githubusercontent.com/26482095/64360141-33313e80-cfe0-11e9-8faa-2fe72a5beed4.png">

Reply:
Hmm OK. I think this is somewhat misleading though as `is_list_like` will return True for `Tuples` as well, so restricting this to `List` might not be totally correct (though I can certainly see where one would think that...)

Does changing this annotation to `Sequence[Callable]` still pass Mypy?

Reply:
It's passes! Although it's necessary to import Sequences in typing and change the concat operation from self.formatters =  fmt[slice] + fmt[slice] to self.formatters = [*fmt[slice],*fmt[slice]] . May I make a commit with theses changes?

Reply:
Hmm what type fails if you keep the code as is but just change annotation? Want to make sure if we change that we have appropriate test coverage

Reply:
I'm not sure I got what you said, but when changing the cast to Sequence[Callable], Mypy calls error on the operand type(+),  like I showed you in the screenshot above. The solution for this was changing the concat form or keeping the cast as a List.

Reply:
So one of two things is wrong here, either:

 1. A Tuple is an allowable value in this block, in which case this would currently fail OR
 2. Our annotations are wrong an a Tuple isn't allowable any way

I *think* the problem here is point 1 above - can you verify that and if so add a test case? Definitely want to get rid of the `cast` unless absolutely necessary. Using `cast` like this can actually hide potential bugs, so need to be careful with it

Reply:
I see what you are saying about hiding bugs, but my understand of this situation is this: the if condition guarantees that we are talking about a self.formatters that is either a list or a tuple and not empty, the exact case where this bug is happening. So here the cast using Sequence would just resolves the Mypy error, because Mypy doesn't actually understand that this is inside that if condition. Therefore, the cast here doesn't seems for me to be able to hide any bug.
That's what I got from this. I maybe got a little confused because it's my first PR so I don't have much experience with pandas-dev. Just trying to get this solved and much appreciate the patience :)

Reply:
Yea totally no worries. Thanks for sticking with this. Instead of `Sequence` if you do `Union[List[Callable], Tuple[Callable, ...]]` does that appease mypy? If so could just go with that

Reply:
Yes it works! Again just like Sequence[Callable] it would be necessary to change concat operation from  self.formatters = fmt[slice] + fmt[slice] to self.formatters = [*fmt[slice],*fmt[slice]] . Now I only need to know if I can go with that or with @simonjayhawkins solution.

Reply:
> it would be necessary to change concat operation from self.formatters = fmt[slice] + fmt[slice] to self.formatters = [*fmt[slice],*fmt[slice]] .

I don't think this should be necessary. changing the code as follows..

```diff
diff --git a/pandas/io/formats/format.py b/pandas/io/formats/format.py
index 5ce4bf6ea..f4bea458f 100644
--- a/pandas/io/formats/format.py
+++ b/pandas/io/formats/format.py
@@ -658,9 +658,10 @@ class DataFrameFormatter(TableFormatter):
                     (frame.iloc[:, :col_num], frame.iloc[:, -col_num:]), axis=1
                 )
                 # truncate formatter
-                if is_list_like(self.formatters) and self.formatters:
-                    truncate_fmt = cast(List[Callable], self.formatters)
-                    self.formatters = truncate_fmt[:col_num] + truncate_fmt[-col_num:]
+                if isinstance(self.formatters, (list, tuple)):
+                    self.formatters = (
+                        self.formatters[:col_num] + self.formatters[-col_num:]
+                    )
             self.tr_col_num = col_num
         if truncate_v:
             # cast here since if truncate_v is True, max_rows_adj is not None
```

mypy will report errors..

```
pandas\io\formats\format.py:663: error: Unsupported operand types for + ("List[Callable[..., Any]]" and "Tuple[Callable[..., Any], ...]")
pandas\io\formats\format.py:663: error: Unsupported operand types for + ("Tuple[Callable[..., Any], ...]" and "List[Callable[..., Any]]")
pandas\io\formats\format.py:663: note: Both left and right operands are unions
```

This looks more like a mypy issue.

Reply:
But in this case why don't we use self.formatters = [*fmt[slice],*fmt[slice]] just to pass in Mypy?

Reply:
the change shouldn't be necessary, but I'm not against you making that change.

I actually like that suggestion, as combined with a change to the isinstance check in `_get_formatter`, would allow us to be more permissive on types accepted for formatters, not just list and tuple for the sequence types.

Reply:
> So one of two things is wrong here, either:
> 
> 1. A Tuple is an allowable value in this block, in which case this would currently fail OR

yes, at present a tuple is allowed
> 2. Our annotations are wrong an a Tuple isn't allowable any way

tuple not documented as acceptable type in the api docs.
 
> I _think_ the problem here is point 1 above - can you verify that and if so add a test case? Definitely want to get rid of the `cast` unless absolutely necessary. Using `cast` like this can actually hide potential bugs, so need to be careful with it

I don't this it is possible to avoid the cast when using is_* functions as mypy cannot perform type narrowing. best solution is to use isinstance checks where possible.

Reply:
@simonjayhawkins do you have a suggestion on how to proceed? I think you know this code better than anyone so would defer to you. Just want to avoid any / all conflicts between annotations, documentation and implementation

Reply:
> Just want to avoid any / all conflicts between annotations, documentation and implementation

makes sense. At the moment we have a conflict between the documentation and the implementation. so that could be confusing matters.

Also, the wording for to_latex is slightly different from to_string and to_html.

they all share the same code.

maybe we should have a  `def _validate_formatters_kwarg(formatters):` since I think the length check is important if this truncation is added.

at the moment, if the list is too short, it raises and if the list is too long it ignores the additional elements.

however, with the slicing from the start and end, weird things might happen.

Reply:
@gabriellm1 in summary..

we want to avoid the cast.

so for now we should probably go with  `if isinstance(self.formatters, (list, tuple)):` instead of `is_list_like(self.formatters) and self.formatters`

this would be consistent with the check in `_get_formatter`

however, you would still need to do `self.formatters = [*fmt[slice],*fmt[slice]] ` to satisfy mypy. I can't find a mypy issue for this case, otherwise we could have added a `# type: ignore` with the issue number.

I think validating the length of the list should be done in this PR, but updating the docs is outside the scope so could be done as a follow-up.

@WillAyd lmk if you disagree.

Reply:
Cast is gone. Now what about length validation?

Reply:
if we only do the length validation (and don't check that the parameter is not a string or that the items are callables) in this PR, then a separate function is probably unnecessary.

so just raise with a message along the lines of "The 'formatters' parameter must be of length equal to the number of columns."?

Reply:
I don't think I got it. The validation would happen in the same place I made the changes? And what did you mean with "raise with a message"?